resgroup: fix startup error when all cpu cores are allocated.

pivotal-ning-yu opened the pull request:
When all the cpu cores are allocated the default cpuset group should
fallback to core 0.  However this fallback logic was only added on
CREATE / ALTER RESOURCE GROUP, but missing in startup logic, an empty cpu
core list "" is set to cgroup and cause a runtime error:

    can't write data to file '/sys/fs/cgroup/cpuset/gpdb/1/cpuset.cpus':
      No space left on device (resgroup-ops-linux.c:916)

Fixed by converting empty cpu bitset to "-1" instead of "".

pivotal-ning-yu commented:
This fix is also needed in 5X branch.

kainwen commented:
+1 LGTM

dujialun commented:
BitsetToCpuset should only be called when Bitmapset is not empty. So this case should be solved in function InitResGroups. Below is a recommended example.
```
        if (gp_resource_group_enable_cgroup_cpuset)
	{
		/*
		 * set default cpuset
		 */
                 if (!bms_is_empty(bmsUnused))
		        BitsetToCpuset(bmsUnused, cpuset, MaxCpuSetLength);
                else
		{
			snprintf(cpuset, MaxCpuSetLength, "%d", defaultCore);
		}
		ResGroupOps_SetCpuSet(DEFAULT_CPUSET_GROUP_ID, cpuset);
	}
```

pivotal-ning-yu commented:
On Tue, Nov 27, 2018 at 12:03 PM Jialun <notifications@github.com> wrote:

> BitsetToCpuset should only be called when Bitmapset is not empty. So this
> case should be solved in function InitResGroups. Below is a recommended
> example.
>
>         if (gp_resource_group_enable_cgroup_cpuset)
> 	{
> 		/*
> 		 * set default cpuset
> 		 */
>                  if (!bms_is_empty(bmsUnused))
> 		        BitsetToCpuset(bmsUnused, cpuset, MaxCpuSetLength);
>                 else
> 		{
> 			snprintf(cpuset, MaxCpuSetLength, "%d", defaultCore);
> 		}
> 		ResGroupOps_SetCpuSet(DEFAULT_CPUSET_GROUP_ID, cpuset);
> 	}
>
>
Makes sense, updated the patch as above.  Also added some comments & checks
in  `BitsetToCpuset()` to describe it requires non-empty bms as input.  The
unit test is also refactored a bit, retired the check on empty bms, switch
to use `assert_string_equal()` instead of `assert_int_equal()` + `strcmp()`.

 —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/6322#issuecomment-441897763>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AYjBHjdghzvQ0BoaClmfMmBUuolpGi4Tks5uzLmXgaJpZM4YyaY2>
> .
>

pivotal-ning-yu commented:
Thanks for the review, merge on master now, will open a PR for 5X later.
